docker-compose.stack.yml

```diff
@@ -0,0 +1,138 @@
+version: "3.7"
+
+x-custom:
+  shared_image: &shared_image
+    image: smirart/postgres-xl:kubernetes
+
+services:
+  pgpool:
+    image: smirart/pgpool:latest
+    ports:
+      - "5432:5432"
+    volumes:
+      - ./pgpool.conf:/etc/pgpool/pgpool.conf
+    restart: always
+    networks:
+      - db-b
+  db-gtm-1:
+    <<: *shared_image
+    environment:
+      - PG_HOST=0.0.0.0
+      - PG_NODE=gtm_1
+      - PG_PORT=6666
+    command: docker-cmd-gtm
+    entrypoint: docker-entrypoint-gtm
+    volumes:
+      - db-gtm-1-data:/var/lib/postgresql
+    networks:
+      - db-a
+    healthcheck:
+      test: ["CMD", "docker-healthcheck-gtm"]
+      interval: 30s
+      timeout: 10s
+      retries: 1
+  db-coord-1:
+    <<: *shared_image
+    environment:
+      - PG_GTM_HOST=db-gtm-1
+      - PG_GTM_PORT=6666
+      - PG_HOST=0.0.0.0
+      - PG_NODE=coord_1
+      - PG_PORT=5432
+      - WAIT_HOSTS=db-gtm-1:6666
+      - WAIT_BEFORE_HOSTS=20
+      - WAIT_HOSTS_TIMEOUT=60
+    command: sh -c "/wait && docker-cmd-coord"
```

Comment:
In general, I'm not usually convinced by wait-for-it type scripts. My reasoning is thus: if things need a special order to start up, that can squash a lot of issues below the service which can appear cluster-wide, but *far* more importantly, it gives a false sense of security, since there's no guarantee that after a container failure or network partition everything will be relaunched in the same order. Even if you're checking health and killing off old things and relaunching the entire container, that still leads to a period where things can be running in an unplanned order cluster-wide, and also, it's very wasteful time-til-recovery-wise, since it requires a full cluster restart just to replace a dead node (e.g. a failed coordinator, especially as Postgres-XL can be so brittle, in my opinion).

This is why there's such a detailed healthcheck system already in this image, including the two-stage bootstrap process for initial initialisation of the cluster, and deep healthchecks for coordinators which even write data into a table despite the performance cost, since in my experience, Postgres-XL can fail very nastily when run in containers, and it's possible to have a coordinator which is live and even working for reads but then ultimately failing for writes. So I guess I'm not sure really what the wait-for-it gives us anyway—I don't use `depends_on` anyway, since it already isn't supported for Docker Swarm via `stack deploy`), and I'm equally concerned it provides a false sense of security. I'd be interested in hearing your reasons further about this.

Reply:
To be honest I've not looked so deeper in that problem. But what I've understood: if you wanna to run async processes you need to synchronise that.

I've thought about two options:

1. To write server-client application who will check container via healthcheck script (because no one knows condition of container better than his healthcheck).
2. Yes we have third option – use orchestrator services like `depends_on` in **docker-compose**, but it hasn't been consisted everywhere (in **docker-stack** f.e.).
3. To use `wait-for-it` and meaning that developers of Postgres-XL care about thus situations (that if something goes wrong – port will be closed). **Yes, it is like take on faith, but it will be simple. Because I've selected this.**

Reply:
1. This sounds like overkill.

2. Yes, but as noted, it's not actually necessary to start the services in this project in any particular order (excepting the bootstrapping requirement). If the GTM is started last, the cluster will converge and restart services as needed already—similar to if there is a networking blip in the night. We wouldn't want to have to wait until we're able to fix it in person, and then restart the whole cluster in a particular order. Preventing this is what the healthchecks try hard to achieve.

3. I can't take it on faith, because I know for a fact that Postgres-XL does not correctly detect all circumstances relating to unhealthy connections. :) e.g. If the network link is interrupted between a coordinator and a datanode, this often isn't even detected. Even with a select-based healthcheck, then it's possible for this to pass—and then have the coordinator fail on the first right (because the underlying links with the datanodes have been severed, or a datanode itself is unhealthy). I agree it would be nice to be able to leave this to Postgres-XL code, however—but this hasn't matched my experience of Postgres-XL issues over the last few years.

Reply:
Okey, I will take one day for the test and give you response about errors related to using / not using `wait-for-it`

Reply:
Thanks. It would be useful to gain an understanding of the specific issues you encountered leading you to add `wait-for-it`, also with any error messages. The healthchecks which are there should already handle the use-case you mention—plus much more. If there is some issue with them, those can probably be fixed without having to change the whole method.